### PR TITLE
Replace TaggedFilesHierarchy with os.walk and implement configure_directory entrypoint

### PR DESCRIPTION
**Pull request by amandarichardsonn**
This PR adds a configure_directory entry point, as well as tests. It also removes TaggedFilesHierarchy and replaces it with os.walk. Finally, the Generator tests have been refactored.

### CODE REVIEW

File: smartsim/_core/generation/generator.py [hunk 1]
```diff
@@ -240,57 +233,50 @@ def _symlink_files(files: t.Union[EntityFiles, None], dest: pathlib.Path) -> Non
     @staticmethod
     def _write_tagged_files(
         files: t.Union[EntityFiles, None],
-        params: t.Mapping[str, str],
+        params: t.Mapping[str, t.Any],
```
Comment (MattToast) [inline smartsim/_core/generation/generator.py:236]: Any reason for the type change here? I think it should be a okay to keep this as a `Mapping[str, str]`. It doesn't make much sense to write anything but a string into a text file imo

File: smartsim/_core/generation/generator.py [hunk 2]
```diff
@@ -240,57 +233,50 @@ def _symlink_files(files: t.Union[EntityFiles, None], dest: pathlib.Path) -> Non
...
-                :param tagged: a TaggedFileHierarchy to be built as a
-                               directory structure
-                """
-                for file in tagged.files:
-                    dst_path = path.join(dest, tagged.base, path.basename(file))
-                    shutil.copyfile(file, dst_path)
+            for file in files.tagged:
+                if os.path.isfile(file):
+                    dst_path = os.path.join(dest, os.path.basename(file))
+                    dst_path = shutil.copyfile(file, dst_path)
```
Comment (MattToast) [inline smartsim/_core/generation/generator.py:254]: Not strictly related to this ticket, but if you want, there should be no reason to manually copy the file only to then configure it place. It looks the the entrypoint is capable of making a new file on the file at the requested location.

File: smartsim/_core/generation/generator.py [hunk 3]
```diff
@@ -240,57 +233,50 @@ def _symlink_files(files: t.Union[EntityFiles, None], dest: pathlib.Path) -> Non
...
-                    mkdir(path.join(dest, tagged.base, path.basename(tagged_dir.base)))
-                    _build_tagged_files(tagged_dir)
-
-            if files.tagged_hierarchy:
-                _build_tagged_files(files.tagged_hierarchy)
-
-            # Pickle the dictionary
+                elif os.path.isdir(file):
+                    dst_path = shutil.copytree(file, dest, dirs_exist_ok=True)
+                    to_write.append(dst_path)
```
Comment (MattToast) [inline smartsim/_core/generation/generator.py:258]: Ditto here: it seems like with a little bit of rework, we can get the entrypoint make a tree on the fly without needing to manually copy everything over only to then configure in place

File: smartsim/_core/generation/generator.py [hunk 4]
```diff
@@ -240,57 +233,50 @@ def _symlink_files(files: t.Union[EntityFiles, None], dest: pathlib.Path) -> Non
...
-        params: t.Mapping[str, str],
+        params: t.Mapping[str, t.Any],
         dest: pathlib.Path,
     ) -> None:
         """Read, configure and write the tagged input files for
            a Job instance. This function specifically deals with the tagged
            files attached to an entity.
 
-        :param app: The Application attached to the Job
+        :param files: The paths to configure
```
Comment (MattToast) [inline smartsim/_core/generation/generator.py:243]: Thanks for fixing these up btw!!

File: tests/test_file_operations.py [hunk 5]
```diff
@@ -583,7 +587,91 @@ def test_configure_not_absolute():
...
+        with pytest.raises(ValueError) as ex:
+            file_operations.configure_directory(ns)
+            assert "param dictionary is empty" in ex.value.args[0]
+    elif error_type == "TypeError":
+        with pytest.raises(TypeError) as ex:
+            file_operations.configure_directory(ns)
+            assert "param dict is not a valid dictionary" in ex.value.args[0]
+    else:
+        file_operations.configure_directory(ns)
+    if error_type == "None":
```
Comment (MattToast) [inline tests/test_file_operations.py:643]: Unnecessary check? Seems like this has to be the case to enter the `else` block here

File: smartsim/_core/entrypoints/file_operations.py [hunk 6]
```diff
@@ -166,39 +182,83 @@ def configure(parsed_args: argparse.Namespace) -> None:
...
+                configure_directory /absolute/file/source/path /absolute/file/dest/path \
+                tag_deliminator param_dict
+
+    /absolute/file/source/path: The tagged directories the search and replace operations
+    to be performed upon
+    /absolute/file/dest/path: The destination for configured directories to be
+    written to.
+    tag_delimiter: tag for the configure_directory operation to search for, defaults to
+        semi-colon e.g. ";"
+    param_dict: A dict of parameter names and values set for the file
```
Comment (juliaputko) [inline smartsim/_core/entrypoints/file_operations.py:238]: change to : "A dict of parameter names and values set for the files in the directory " or something similar

File: tests/test_generator.py [hunk 7]
```diff
@@ -213,19 +234,85 @@ def test_generate_configure(fileutils, wlmutils, generator_instance):
...
-
-    # Call Generator.generate_job
-    path, _, _ = generator_instance.generate_job(job, 0)
-    # Retrieve the list of configured files in the test directory
-    configured_files = sorted(glob(str(path) + "/*"))
-    # Use filecmp.cmp to check that the corresponding files are equal
+    generator_instance._write_tagged_files(
+        files=files, params=param_set, dest=generator_instance.root
+    )
+    configured_files = sorted(glob(os.path.join(generator_instance.root) + "/*"))
```
Comment (MattToast) [inline tests/test_generator.py:240]: Nit: Did you mean
```suggestion
    configured_files = sorted(glob(os.path.join(generator_instance.root, "*")))
```
?

**Review by juliaputko — APPROVED** [review]
1 minor comment, otherwise lgtm!!

**Review by MattToast — CHANGES_REQUESTED** [review]
Just some initial feedback while the CI gets worked out. So far looks good!

File: smartsim/_core/entrypoints/file_operations.py [hunk 8]
```diff
@@ -166,39 +182,83 @@ def configure(parsed_args: argparse.Namespace) -> None:
...
+    substitutions = tuple(
+        _make_substitution(k, v, tag_delimiter) for k, v in param_dict.items()
+    )
+    try:
+        # Set the lines to iterate over
+        with open(parsed_args.source, "r+", encoding="utf-8") as file_stream:
+            try:
+                lines = [_replace_tags_in(line, substitutions) for line in file_stream]
+            except IOError as e:
+                raise e from None
```
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply smartsim/_core/entrypoints/file_operations.py:208]: I actually wanted to ask about this, is it worth catching IOError and OSErrors during an open call?

[the remark this answers is not in the document]
Comment (MattToast) [reply smartsim/_core/entrypoints/file_operations.py:208]: If we are going to handle the error in someway, I'm never going to discourage it, but if we are just going to re-raise without doing anything else, I say save the indentation.

File: smartsim/_core/entrypoints/file_operations.py (hunk 6 above)
Comment (amandarichardsonn) [reply smartsim/_core/entrypoints/file_operations.py:238]: thank you!

File: smartsim/_core/entrypoints/file_operations.py [hunk 9]
```diff
@@ -62,6 +62,22 @@ def _make_substitution(
...
+    This function takes a base64-encoded string representation of a dictionary,
+    decodes it, and then deserializes it using pickle. It performs validation
+    to ensure the resulting object is a non-empty dictionary.
+    """
+    decoded_dict = base64.b64decode(param_dict)
+    deserialized_dict = pickle.loads(decoded_dict)
+    if not deserialized_dict:
+        raise ValueError("param dictionary is empty")
+    if not isinstance(deserialized_dict, dict):
+        raise TypeError("param dict is not a valid dictionary")
```
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply smartsim/_core/entrypoints/file_operations.py:77]: very true... thank you!

File: smartsim/_core/entrypoints/file_operations.py [hunk 10]
```diff
@@ -166,39 +182,83 @@ def configure(parsed_args: argparse.Namespace) -> None:
...
+                    os.path.join(dirpath, file_name), "r+", encoding="utf-8"
+                ) as file_stream:
+                    lines = [
+                        _replace_tags_in(line, substitutions) for line in file_stream
+                    ]
+                # write configured file to destination specified
+                with open(
+                    os.path.join(dirpath, file_name), "w+", encoding="utf-8"
+                ) as file_stream:
+                    file_stream.writelines(lines)
```
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply smartsim/_core/entrypoints/file_operations.py:261]: thank you!

**Review by mellis13 — CHANGES_REQUESTED** [review]
Just some general comments to get the review process started.

File: smartsim/_core/generation/generator.py [hunk 11]
```diff
@@ -136,23 +164,19 @@ def generate_job(
...
             dt_string = datetime.now().strftime("%d/%m/%Y %H:%M:%S")
             log_file.write(f"Generation start date and time: {dt_string}\n")
 
-        # Create output files
-        out_file, err_file = self._output_files(log_path, job.entity.name)
+        out_file = self._out_file(log_path, job.entity.name)
+        err_file = self._err_file(log_path, job.entity.name)
 
-        # Perform file system operations on attached files
         self._build_operations(job, job_path)
```
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply smartsim/_core/generation/generator.py:187]: Ah yes, will post rn

File: smartsim/_core/generation/generator.py [hunk 12]
```diff
@@ -28,206 +28,276 @@
...
+
 class Generator:
-    """The primary job of the Generator is to create the directory and file structure
-    for a SmartSim Job. The Generator is also responsible for writing and configuring
-    files into the Job directory.
-    """
+    """The primary responsibility of the Generator class is to create the directory structure
+    for a SmartSim Job and to build and execute file operation commands."""
+
+    run = "run"
```
Comment (mellis13) [inline smartsim/_core/generation/generator.py:62]: Would ``run_directory`` be a better variable name?

File: smartsim/_core/generation/generator.py [hunk 13]
```diff
@@ -28,206 +28,276 @@
...
+    """The name of the directory where run-related files are stored."""
+    log = "log"
```
Comment (mellis13) [inline smartsim/_core/generation/generator.py:64]: Would ``log_directory`` be a better variable name?

File: smartsim/_core/generation/generator.py [hunk 14]
```diff
@@ -28,206 +28,276 @@
...
+    """The name of the directory where log files are stored."""
 
     def __init__(self, root: pathlib.Path) -> None:
         """Initialize a Generator object
 
-        The class handles symlinking, copying, and configuration of files
-        associated with a Jobs entity. Additionally, it writes entity parameters
-        used for the specific run into the "smartsim_params.txt" settings file within
-        the Jobs log folder.
+        The Generator class constructs a Jobs directory structure, including:
```
Comment (mellis13) [inline smartsim/_core/generation/generator.py:70]: Jobs -> Job's

File: smartsim/_core/generation/generator.py [hunk 15]
```diff
@@ -28,206 +28,276 @@
...
+
+        - The run and log directories
+        - Output and error files
+        - The "smartsim_params.txt" settings file
+
+        Additionally, it manages symlinking, copying, and configuring files associated
+        with a Jobs entity.
```
Comment (mellis13) [inline smartsim/_core/generation/generator.py:77]: Jobs -> Job's

File: smartsim/_core/generation/generator.py [hunk 16]
```diff
@@ -28,206 +28,276 @@
...
+
+        :param root: The Jobs base path
```
Comment (mellis13) [inline smartsim/_core/generation/generator.py:79]: The Jobs base path -> Job base path

File: smartsim/_core/generation/generator.py [hunk 17]
```diff
@@ -28,206 +28,276 @@
...
         """
         self.root = root
         """The root path under which to generate files"""
 
-    def _generate_job_root(self, job: Job, job_index: int) -> pathlib.Path:
-        """Generates the root directory for a specific job instance.
+    def _build_job_base_path(self, job: Job, job_index: int) -> pathlib.Path:
+        """Build and return a Jobs base directory. The path is created by combining the
```
Comment (mellis13) [inline smartsim/_core/generation/generator.py:85]: Jobs -> Job's

File: smartsim/_core/generation/generator.py [hunk 18]
```diff
@@ -28,206 +28,276 @@
...
         """
         job_type = f"{job.__class__.__name__.lower()}s"
         job_path = self.root / f"{job_type}/{job.name}-{job_index}"
         return pathlib.Path(job_path)
 
-    def _generate_run_path(self, job: Job, job_index: int) -> pathlib.Path:
-        """Generates the path for the "run" directory within the root directory
-        of a specific Job instance.
+    def _build_job_run_path(self, job: Job, job_index: int) -> pathlib.Path:
+        """Build and return a Jobs run directory. The path is formed by combining
```
Comment (mellis13) [inline smartsim/_core/generation/generator.py:99]: Jobs -> Job's

File: smartsim/_core/generation/generator.py [hunk 19]
```diff
@@ -28,206 +28,276 @@
...
-        :param job: The Job instance for which the path is generated.
-        :param job_index: The index of the Job instance (used for naming).
-        :returns: The path to the "log" directory for the Job instance.
+        :param job: The job object
+        :param job_index: The index of the Job
+        :returns: The built file path for the Job run folder
         """
-        path = self._generate_job_root(job, job_index) / "log"
+        path = self._build_job_base_path(job, job_index) / self.log
         path.mkdir(exist_ok=False, parents=True)
```
Comment (mellis13) [inline smartsim/_core/generation/generator.py:121]: Some of these ``_build`` functions call ``mkdir`` and some don't.  I woudl proposed that the ``_build`` functions only build the ``pathlib.Path`` directories and the ``mkdir`` are only called in the ``execute_command`` function.

**Review by mellis13 — CHANGES_REQUESTED** [review]
Just a couple of comments/questions.

File: smartsim/_core/generation/generator.py (hunk 12 above)
Comment (amandarichardsonn) [reply smartsim/_core/generation/generator.py:62]: ah yes, run is vague

File: smartsim/_core/generation/generator.py (hunk 18 above)
Comment (amandarichardsonn) [reply smartsim/_core/generation/generator.py:99]: ah my bad on all those

File: smartsim/_core/generation/generator.py [hunk 20]
```diff
@@ -236,67 +314,51 @@ def _symlink_files(files: t.Union[EntityFiles, None], dest: pathlib.Path) -> Non
...
+                        tag_delimiter,
                         encoded_dict,
                     ]
                 )
+                cmd_list.commands.append(cmd)
+        return cmd_list
 
-            # TODO address in ticket 723
-            # self._log_params(entity, files_to_params)
+        # TODO address in ticket 723
```
Comment (mellis13) [inline smartsim/_core/generation/generator.py:360]: Can we add this code down here to ticket 723 and delete?

**Review by mellis13 — APPROVED** [review]
Only one small item but otherwise I think this is ready to merge and other items can be addressed as the user API for attaching files is updating.  Thanks!

File: smartsim/_core/generation/generator.py (hunk 20 above)
Comment (amandarichardsonn) [reply smartsim/_core/generation/generator.py:360]: yes!

**Pull request merged**